docs/rfc/active/000-change-processor/README.md

```diff
@@ -0,0 +1,111 @@
+
+### Proposed implementation
+
+1. On metadata-service start up, use a ServiceLoader to load a jar from a specified location (defined in config). In the
+   future this could be extended to load in jars at runtime.
+2. The user defined jar would contain a class that implements a [interface](#contract-example) that the metadata service
+   can understand.
+3. Attach an annotation to the processor that specifies what entity/aspect the processor should apply to, and whether
+   the processor is run before or after it's applied to the datastore.
+4. On each update call pass the previous aspect version and the proposed new version into a processor that runs it
+   through all registered processors.
+5. If a validation fails bubble up the details to the caller.
+
+If no custom jar is detected on start up keep the existing noop behaviour.
+
+### Developer Lifecycle Example
+
+Suppose we require the ability to ignore metadata updates where a property contains a specific string. The following
+steps indicates the development lifecycle:
+
+1. Create a new project that will contain the custom validation logic.
+2. Add a dependency to the DataHub metadata-io project that contains the expected contracts.
+3. Implement the class and write the change processors.
+4. Build and deploy the jar to a location the running metadata-service will have access to.
+5. Restart the metadata-service and provide the path to the new jar via config.
+6. All new updates will pass through the new code path checking whether that aspect change contains the disallowed
+   string.
+
+### Contracts Example
+[Change processor interface](../../../../metadata-io/src/main/java/com/linkedin/metadata/changeprocessor/ChangeProcessor.java): Interface that clients will develop their custom processor logic against. 
+
+[Change processor annotation](../../../../metadata-io/src/main/java/com/linkedin/metadata/changeprocessor/ChangeProcessorScope.java): Allows processors to target specific entity aspect combinations.
+
+[Change stream processor](../../../../metadata-io/src/main/java/com/linkedin/metadata/changeprocessor/ChangeStreamProcessor.java): Responsible for running all the registered processors
+
+## Drawbacks
+
+* Opens up ability to add unwanted side effects that DataHub cannot control. 
+* Difficult to see what validation is being applied to what aspect at runtime.
+* Deployment of DataHub would be more complicated than it is currently, before starting the metadata service the jar
+  must exist in the expected location.
+
+## Alternatives
+
+* Modify DataHub to provide workflow functionality, where users can upload
+  a [BPMN](https://en.wikipedia.org/wiki/Business_Process_Model_and_Notation) diagram into DataHub and offer similar
+  functionality to workflow engines like Camunda. Downsides include introducing much more complexity into DataHub.
+
+* Instead of loading validation logic by jars, create a validation service that DataHub calls out to. Requires DataHub
```

Comment:
Yeah - if we add these interfaces we having something much more general purpose - the implementations can of course decide to make outbound calls if required